File: android/app/src/main/java/ie/oireachtas/explorer/data/collections/PublicCollections.kt
package ie.oireachtas.explorer.data.collections

import com.squareup.moshi.JsonClass
import com.squareup.moshi.Moshi
import com.squareup.moshi.kotlin.reflect.KotlinJsonAdapterFactory
import ie.oireachtas.explorer.BuildConfig
import ie.oireachtas.explorer.data.saved.SavedItem
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody

/**
 * Public, share-by-link research collections — backed by the same
 * Cloudflare Worker that the web app talks to. Mirrors the contract in
 * src/api/publicCollections.ts.
 *
 * The base URL is wired via the Gradle `workerBaseUrl` property; an
 * empty value disables the feature at runtime so the UI can fall back
 * to a "not configured" notice (matching the web's behaviour).
 */

@JsonClass(generateAdapter = true)
data class PublicCollection(
    val slug: String,
    val title: String,
    val description: String? = null,
    val createdAt: String,
    val itemCount: Int,
    val items: List<SavedItem>,
)

@JsonClass(generateAdapter = true)
data class PublishCollectionInput(
    val title: String,
    val description: String? = null,
    val items: List<SavedItem>,
)

class PublicCollectionException(message: String) : RuntimeException(message)

object PublicCollections {
    private val moshi = Moshi.Builder().add(KotlinJsonAdapterFactory()).build()
    private val collectionAdapter = moshi.adapter(PublicCollection::class.java)
    private val inputAdapter = moshi.adapter(PublishCollectionInput::class.java)
    private val client = OkHttpClient()

    fun isEnabled(): Boolean = BuildConfig.WORKER_BASE_URL.isNotBlank()

    private fun base(): String {
        val raw = BuildConfig.WORKER_BASE_URL
        if (raw.isBlank()) {
            throw PublicCollectionException("Public collections need a configured Cloudflare Worker URL.")
        }
        return raw.trimEnd('/')
    }

    suspend fun publish(input: PublishCollectionInput): PublicCollection = withContext(Dispatchers.IO) {
        val body = inputAdapter.toJson(input).toRequestBody("application/json".toMediaType())
        val req = Request.Builder().url("${base()}/collections").post(body).build()
        execute(req)
    }

    suspend fun fetch(slug: String): PublicCollection = withContext(Dispatchers.IO) {
        val req = Request.Builder().url("${base()}/collections/$slug").get().build()
        execute(req)
    }

    private fun execute(request: Request): PublicCollection {
        client.newCall(request).execute().use { response ->
            val body = response.body?.string().orEmpty()
            if (!response.isSuccessful) {
                throw PublicCollectionException("Request failed (${response.code}). $body".trim())
            }
            return collectionAdapter.fromJson(body)
                ?: throw PublicCollectionException("Empty response from worker.")
        }
    }
}

File: src/api/publicCollections.ts
import type { SavedItem } from '../utils/savedItems';

const WORKER_API_BASE = (import.meta.env.VITE_TRANSCRIPT_API_BASE as string | undefined)?.replace(/\/+$/, '') ?? '';

export interface PublicResearchCollection {
  slug: string;
  title: string;
  description?: string;
  createdAt: string;
  itemCount: number;
  items: SavedItem[];
}

export interface PublishCollectionInput {
  title: string;
  description?: string;
  items: SavedItem[];
}

export class PublicCollectionError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'PublicCollectionError';
  }
}

function ensureWorkerBase(): string {
  if (!WORKER_API_BASE) {
    throw new PublicCollectionError('Public collections need a configured Cloudflare Worker URL.');
  }
  return WORKER_API_BASE;
}

async function readError(response: Response): Promise<string> {
  try {
    const data: unknown = await response.json();
    if (data && typeof data === 'object' && 'error' in data && typeof data.error === 'string') {
      return data.error;
    }
  } catch {
    // ignore
  }
  return `Request failed (${response.status}).`;
}

export function isPublicCollectionsEnabled(): boolean {
  return Boolean(WORKER_API_BASE);
}

export async function publishPublicCollection(input: PublishCollectionInput): Promise<PublicResearchCollection> {
  const base = ensureWorkerBase();
  const response = await fetch(`${base}/collections`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(input),
  });

  if (!response.ok) {
    throw new PublicCollectionError(await readError(response));
  }

  return await response.json() as PublicResearchCollection;
}

export async function fetchPublicCollection(slug: string, signal?: AbortSignal): Promise<PublicResearchCollection> {
  const base = ensureWorkerBase();
  const response = await fetch(`${base}/collections/${encodeURIComponent(slug)}`, { signal });

  if (!response.ok) {
    throw new PublicCollectionError(await readError(response));
  }

  return await response.json() as PublicResearchCollection;
}
